### ISSUE TITLE
Crash during ALTER that renames a column and adds a new column AFTER the renamed column

### ISSUE THREAD

**Issue by bryantraywick** [zendesk/maxwell#1216]
Maxwell crashes on the following `ALTER`:

    ALTER TABLE `production`.`_tasks_new` ADD COLUMN taskable_type varchar(255) DEFAULT 'Item' AFTER taskable_id, DROP INDEX index_tasks_on_item_id, CHANGE COLUMN item_id taskable_id int(11) DEFAULT NULL, ADD INDEX index_tasks_on_taskable_id_and_taskable_type (taskable_id, taskable_type)

Here are the relevant Maxwell logs:

    12:33:14,175 INFO  AbstractSchemaStore - storing schema @Position[BinlogPosition[mariadb-bin.013329:7123672], lastHeartbeat=1550489706636] after applying "CREATE TABLE `production`.`_tasks_new` (   `id` int(11) NOT NULL AUTO_INCREMENT,   `task_type` varchar(255) NOT NULL,   `item_id` int(11) NOT NULL,   `assignee_id` int(11) DEFAULT NULL,   `started_at` datetime DEFAULT NULL,   `ended_at` datetime DEFAULT NULL,   `created_at` datetime NOT NULL,   `updated_at` datetime NOT NULL,   `updated_by_id` int(11) DEFAULT NULL,   `created_by_id` int(11) DEFAULT NULL,   PRIMARY KEY (`id`),   KEY `index_tasks_on_item_id` (`item_id`),   KEY `index_tasks_on_assignee_id` (`assignee_id`),   KEY `index_tasks_on_task_type` (`task_type`),   KEY `index_tasks_on_started_at` (`started_at`),   KEY `index_tasks_on_ended_at` (`ended_at`),   KEY `index_tasks_on_updated_by_id` (`updated_by_id`),   KEY `index_tasks_on_created_by_id` (`created_by_id`),     CONSTRAINT `_fk_rails_0016c50613` FOREIGN KEY (`assignee_id`) REFERENCES `users` (`id`),     CONSTRAINT `_fk_rails_8578e0e40d` FOREIGN KEY (`updated_by_id`) REFERENCES `users` (`id`),     CONSTRAINT `_fk_rails_a362a150d3` FOREIGN KEY (`created_by_id`) REFERENCES `users` (`id`) ) ENGINE=InnoDB AUTO_INCREMENT=953748 DEFAULT CHARSET=utf8" to production, new schema id is 829
    12:33:14,231 ERROR AbstractSchemaStore - Error on bin log position Position[BinlogPosition[mariadb-bin.013329:7124911], lastHeartbeat=1550489706636]
    com.zendesk.maxwell.schema.ddl.InvalidSchemaError: Could not find column taskable_id (needed in AFTER statement)
    	at com.zendesk.maxwell.schema.ddl.ColumnPosition.index(ColumnPosition.java:28)
    	at com.zendesk.maxwell.schema.ddl.AddColumnMod.apply(AddColumnMod.java:18)
    	at com.zendesk.maxwell.schema.ddl.TableAlter.resolve(TableAlter.java:52)
    	at com.zendesk.maxwell.schema.ddl.TableAlter.resolve(TableAlter.java:12)
    	at com.zendesk.maxwell.schema.AbstractSchemaStore.resolveSQL(AbstractSchemaStore.java:58)
    	at com.zendesk.maxwell.schema.MysqlSchemaStore.processSQL(MysqlSchemaStore.java:104)
    	at com.zendesk.maxwell.replication.BinlogConnectorReplicator.processQueryEvent(BinlogConnectorReplicator.java:244)
    	at com.zendesk.maxwell.replication.BinlogConnectorReplicator.processQueryEvent(BinlogConnectorReplicator.java:262)
    	at com.zendesk.maxwell.replication.BinlogConnectorReplicator.getRow(BinlogConnectorReplicator.java:492)
    	at com.zendesk.maxwell.replication.BinlogConnectorReplicator.work(BinlogConnectorReplicator.java:151)
    	at com.zendesk.maxwell.util.RunLoopProcess.runLoop(RunLoopProcess.java:27)
    	at com.zendesk.maxwell.Maxwell.startInner(Maxwell.java:224)
    	at com.zendesk.maxwell.Maxwell.start(Maxwell.java:156)
    	at com.zendesk.maxwell.Maxwell.main(Maxwell.java:245)
    12:33:14,332 INFO  BinlogConnectorLifecycleListener - Binlog disconnected.
    12:33:14,349 INFO  TaskManager - Stopping 4 tasks
    12:33:14,349 ERROR TaskManager - cause:
    com.zendesk.maxwell.schema.ddl.InvalidSchemaError: Could not find column taskable_id (needed in AFTER statement)
    	at com.zendesk.maxwell.schema.ddl.ColumnPosition.index(ColumnPosition.java:28) ~[maxwell-1.19.1.jar:1.19.1]
    	at com.zendesk.maxwell.schema.ddl.AddColumnMod.apply(AddColumnMod.java:18) ~[maxwell-1.19.1.jar:1.19.1]
    	at com.zendesk.maxwell.schema.ddl.TableAlter.resolve(TableAlter.java:52) ~[maxwell-1.19.1.jar:1.19.1]
    	at com.zendesk.maxwell.schema.ddl.TableAlter.resolve(TableAlter.java:12) ~[maxwell-1.19.1.jar:1.19.1]
    	at com.zendesk.maxwell.schema.AbstractSchemaStore.resolveSQL(AbstractSchemaStore.java:58) ~[maxwell-1.19.1.jar:1.19.1]
    	at com.zendesk.maxwell.schema.MysqlSchemaStore.processSQL(MysqlSchemaStore.java:104) ~[maxwell-1.19.1.jar:1.19.1]
    	at com.zendesk.maxwell.replication.BinlogConnectorReplicator.processQueryEvent(BinlogConnectorReplicator.java:244) ~[maxwell-1.19.1.jar:1.19.1]
    	at com.zendesk.maxwell.replication.BinlogConnectorReplicator.processQueryEvent(BinlogConnectorReplicator.java:262) ~[maxwell-1.19.1.jar:1.19.1]
    	at com.zendesk.maxwell.replication.BinlogConnectorReplicator.getRow(BinlogConnectorReplicator.java:492) ~[maxwell-1.19.1.jar:1.19.1]
    	at com.zendesk.maxwell.replication.BinlogConnectorReplicator.work(BinlogConnectorReplicator.java:151) ~[maxwell-1.19.1.jar:1.19.1]
    	at com.zendesk.maxwell.util.RunLoopProcess.runLoop(RunLoopProcess.java:27) ~[maxwell-1.19.1.jar:1.19.1]
    	at com.zendesk.maxwell.Maxwell.startInner(Maxwell.java:224) ~[maxwell-1.19.1.jar:1.19.1]
    	at com.zendesk.maxwell.Maxwell.start(Maxwell.java:156) ~[maxwell-1.19.1.jar:1.19.1]
    	at com.zendesk.maxwell.Maxwell.main(Maxwell.java:245) ~[maxwell-1.19.1.jar:1.19.1]
    12:33:14,355 INFO  TaskManager - Stopping: com.zendesk.maxwell.schema.PositionStoreThread@75455fe6
    12:33:14,355 INFO  StoppableTaskState - com.zendesk.maxwell.schema.PositionStoreThread requestStop() called (in state: RUNNING)

We use `pt-online-schema-change` but I believe this issue would occur for a normal `ALTER` to the original table. The issue appears to be that Maxwell isn't yet aware of the new column `taskable_id` that is added in the same `ALTER TABLE`, and is therefore unable to process the `AFTER taskable_id` portion of the query.

I was able to workaround the problem by manually updating the Maxwell binlog position to just after `pt-online-schema-change`'s `RENAME TABLE` and trigger drops, and following the advice in #487 I truncated `tables`, `schemas`, `databases`, and `columns` so that Maxwell would refresh the schema. But this certainly looks like a bug with Maxwell's handling of this type of scenario. I have advised everyone to break such `ALTER TABLE`s into two statements going forward to avoid this problem: one that adds the new column (`taskable_id` in this case) and one that adds the new column `AFTER` that new column.

I don't know if the problem is isolated to just cases where a column is renamed and then a new column is added after the renamed column. Or if the same problem would occur if two brand new columns were added in the same `ALTER TABLE` with one of them specified to be `AFTER` the other. Also, upgrading to 1.19.7 while troubleshooting did not resolve this problem.

**Comment by osheroff** [comment]
nasty.  ok, will look.

**Comment by osheroff** [comment]
you definitely found an odd little corner of mysql; your DDL references a renamed column before it's renamed in the alter.  FWIW:

```
create table t ( i int );
# this works
alter table t change column i j int, add column n int after j;
# this crashes maxwell
alter table t add column n int after j, change column i j int;
```

AFAICT this only applies to `CHANGE` and `RENAME COLUMN` stuff, if I try to add a column that references another column out of order mysql properly denies me:

```
mysql> alter table t add column n int after j, add column j int;
ERROR 1054 (42S22): Unknown column 'j' in 't'
```

This should be a reasonably straightforward fix.

**Comment by osheroff** [comment]
oh good lord, mysql is really way too damn permissive with this:

```
create table t ( a int, b varchar(255));
 alter table t change column a a int after b_gets_renamed, change column b b_gets_renamed varchar(255);
```

which seems to say that you can always reference the new names of columns, it's not just a matter of processing column changes first... drrrrrfffffff......

**Comment by osheroff** [comment]
fixed in v1.20.0
